pkg/controller/worker/state_actuator.go

```diff
@@ -0,0 +1,154 @@
+func (a *genericStateActuator) Reconcile(ctx context.Context, worker *extensionsv1alpha1.Worker) error {
+	copyOfWorker := worker.DeepCopy()
+	if err := a.updateWorkerState(ctx, copyOfWorker); err != nil {
+		return errors.Wrapf(err, "failed to update the state in worker status")
+	}
+
+	return nil
+}
+
+func (a *genericStateActuator) updateWorkerState(ctx context.Context, worker *extensionsv1alpha1.Worker) error {
+	state, err := a.getWorkerState(ctx, worker)
+	if err != nil {
+		return err
+	}
+
+	err = extensionscontroller.TryUpdateStatus(ctx, retry.DefaultBackoff, a.client, worker, func() error {
+		worker.Status.State = &runtime.RawExtension{Raw: state}
+		return nil
+	})
+	return err
+}
+
+func (a *genericStateActuator) getWorkerState(ctx context.Context, worker *extensionsv1alpha1.Worker) ([]byte, error) {
+	existingMachineDeployments := &machinev1alpha1.MachineDeploymentList{}
+	if err := a.client.List(ctx, existingMachineDeployments, client.InNamespace(worker.Namespace)); err != nil {
+		return nil, err
+	}
+
+	var machineDeploymentNames []string
+	for _, machineDeployment := range existingMachineDeployments.Items {
+		machineDeploymentNames = append(machineDeploymentNames, machineDeployment.Name)
+	}
+
+	existingMachineSets := &machinev1alpha1.MachineSetList{}
+	if err := a.client.List(ctx, existingMachineSets, client.InNamespace(worker.Namespace)); err != nil {
+		return nil, err
+	}
+
+	machineSets := make(map[string]*machinev1alpha1.MachineSet)
+	for index, machineSet := range existingMachineSets.Items {
+		for _, referant := range machineSet.OwnerReferences {
+			if referant.Kind == "MachineDeployment" {
+				machineSets[referant.Name] = &existingMachineSets.Items[index]
+			}
+		}
+	}
+
+	existingMachines := &machinev1alpha1.MachineList{}
+	if err := a.client.List(ctx, existingMachines, client.InNamespace(worker.Namespace)); err != nil {
+		return nil, err
+	}
+
+	machines := make(map[string][]*machinev1alpha1.Machine)
+	for index, machine := range existingMachines.Items {
+		for _, referant := range machine.OwnerReferences {
+			if referant.Kind == "MachineSet" {
+				machines[referant.Name] = append(machines[referant.Name], &existingMachines.Items[index])
+			}
+		}
+	}
+
+	workerState := make(map[string]*MachineDeploymentState)
+	for _, deploymentName := range machineDeploymentNames {
+		machineDeploymentState := &MachineDeploymentState{}
+
+		machineSet, ok := machineSets[deploymentName]
+		if !ok {
+			continue
+		}
+
+		//remove redundant data from the machine set
+		machineSet.ObjectMeta = metav1.ObjectMeta{
+			Name:        machineSet.Name,
+			Namespace:   machineSet.Namespace,
+			Annotations: machineSet.Annotations,
+			Labels:      machineSet.Labels,
+		}
+		machineSet.OwnerReferences = nil
+
+		machineDeploymentState.MachineSet = &runtime.RawExtension{Object: machineSet}
+
+		currentMachines := machines[machineSet.Name]
+		if len(currentMachines) <= 0 {
+			continue
```

Comment:
This check is not needed. If the length is zero then your `for` loop will simply do nothing. Actually, `continue`ing here is even a bug, or? The line `150` would be skipped which is undesired, or?

Reply:
Do I have to save state of a `MachineDeployment` which has no machines?
How to tell json.Marshal(workerState) to ignore keys which have empty `MachineDeploymentState`?

Reply:
Hm, okay, I understand. No, you don't have to save the state, of course, and if you have problems with `json.Marshal` then it's fine to `continue` here. Thanks.